// vue-styleguidist/lib/scripts/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Config", {
  enumerable: true,
  get: function get() {
    return _StyleGuide.StyleguidistConfig;
  }
});
exports["default"] = _default;

require("react-styleguidist/lib/scripts/utils/ensureWebpack");

var _logger = _interopRequireDefault(require("react-styleguidist/lib/scripts/logger"));

var _StyleGuide = require("../types/StyleGuide");

var _build = _interopRequireDefault(require("./build"));

var _server2 = _interopRequireDefault(require("./server"));

var _makeWebpackConfig2 = _interopRequireDefault(require("./make-webpack-config"));

var _config = _interopRequireDefault(require("./config"));

var binutils = _interopRequireWildcard(require("./binutils"));

var _isPromise = _interopRequireDefault(require("./utils/isPromise"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// Make sure user has webpack installed

/**
 * Initialize Vue Styleguide API.
 *
 * @param {object} [config] Styleguidist config.
 * @param {function} [updateConfig] update config post resolution
 * @returns {object} API.
 */
function _default(config, updateConfig) {
  var configInternal = (0, _config["default"])(config, function (cfg) {
    (0, _logger["default"])(cfg.logger, cfg.verbose, {});

    if (typeof updateConfig === 'function') {
      updateConfig(cfg);
    }

    return cfg;
  });

  if ((0, _isPromise["default"])(configInternal)) {
    return configInternal.then(function (conf) {
      return exportBuildUtils(conf);
    })["catch"](function (e) {
      throw e;
    });
  } else {
    return exportBuildUtils(configInternal);
  }
}

function exportBuildUtils(config) {
  return {
    build: function build(callback) {
      return (0, _build["default"])(config, function (err, stats) {
        return callback(err, config, stats);
      });
    },
    server: function server(callback) {
      return (0, _server2["default"])(config, function (err) {
        return callback(err, config);
      });
    },
    makeWebpackConfig: function makeWebpackConfig(env) {
      return (0, _makeWebpackConfig2["default"])(config, env || 'production');
    },
    binutils: {
      server: function server(open) {
        return binutils.commandServer(config, open);
      },
      build: function build() {
        return binutils.commandBuild(config);
      }
    }
  };
}
